### PR TITLE
Add integer array indexing on one axis for __getitem__

### PR DESCRIPTION
**Pull request by yuyu2172**
The PR change `ndarray.__getitem__` to accept `ndarray` or `numpy.ndarray` in the argument. It checks for the `dtype` of arrays in the argument, and raise Error when necessary. Currently, only integer arrays are supported. 

The modification simply integrates `take` and `__getitem__`. With this PR, `a[:, idx]` returns same value as `a.take(idx, axis=1)`.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1016,17 +1017,30 @@ cdef class ndarray:
...
 
-        for s in slices:
-            if isinstance(s, ndarray):
-                raise ValueError('Advanced indexing is not supported')
+        # Check if advanced is true and if there are multiple integer indexing
+        advanced = False
+        axis = None
+        for i, s in enumerate(slices):
+            if isinstance(s, (numpy.ndarray, ndarray)):
+                if s.dtype in [numpy.int64, numpy.int32,
```
Comment (okuta) [inline cupy/core/core.pyx]: How about using `issubclass(s.dtype.type, numpy.integer)`?

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1039,52 +1053,64 @@ cdef class ndarray:
...
-                if ind < 0:
-                    ind += self._shape[j]
-                if not (0 <= ind < self._shape[j]):
-                    msg = ('Index %s is out of bounds for axis %s with size %s'
-                           % (s, j, self._shape[j]))
-                    raise IndexError(msg)
-                offset += ind * self._strides[j]
-                j += 1
+        if advanced:
+            if (axis is not None and
```
Comment (okuta) [inline cupy/core/core.pyx]: How about following code?
```
if axis is not None:
    noneSlice = slice(None)
    flag = True
    for i, s in enumerate(slices):
        if i != axis and s != slice:
            flag = False
            break
    if flag:
        return self.take(slices[axis], axis)
raise NotImplementedError(
    'Advanced indexing with 2 or more arrays is not supported')
```

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -1039,52 +1053,64 @@ cdef class ndarray:
...
             else:
-                raise TypeError('Invalid index type: %s' % type(slices[i]))
+                raise NotImplementedError('Adv indexing with 2 or ' +
+                                          'more arrays is not supported')
 
-        v = self.view()
-        v.data = self.data + offset
-        v._set_shape_and_strides(shape, strides)
-        return v
+        else:
```
Comment (okuta) [inline cupy/core/core.pyx]: Please remove this `else`.

**Comment by yuyu2172** [comment]
Sorry. I pushed rebased code, and the commit log got messed up.
The first three commits (95590d8 cf8a700 81223b1) were reviewed.

Some additional commits (c5900aa, 94df0b3, fa8b1c5) are all about handling special cases.

File: cupy/core/core.pyx (hunk 1 above)
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thank you!

File: cupy/core/core.pyx (hunk 2 above)
Comment (yuyu2172) [inline cupy/core/core.pyx]: That is much better.

File: cupy/core/core.pyx (hunk 3 above)
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thank you!

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -1039,6 +1050,42 @@ cdef class ndarray:
...
+        # Check if advanced is true and if there are multiple integer indexing
+        advanced = False
+        axis = None
+        for i, s in enumerate(slices):
+            if isinstance(s, list):
+                s = numpy.array(s)
+                slices[i] = s
+            if isinstance(s, (numpy.ndarray, ndarray)):
+                if issubclass(s.dtype.type, numpy.integer):
+                    if advanced:
```
Comment (yuyu2172) [inline cupy/core/core.pyx]: Thank you for commenting.

`advanced` is `True` when there are two or more arrays in `slices`.
In that case, `NotImplementedError` at line 1086 will be raised.

Comment (okuta) [inline cupy/core/core.pyx]: Oh, sorry. This code is in the loop. I understood.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**